### ISSUE TITLE
Add ability to iterate through flags

### ISSUE THREAD

**Issue by gsingh93** [bitflags/bitflags#28]
Iterating over each each possible flag value would be a cool feature. I'm currently making an array with each constant value and then iterating over that.

**Comment by RReverser** [comment]
@alexcrichton Is this something that would be accepted in the library?

**Comment by alexcrichton** [comment]
Seems plausible!

**Comment by djg** [comment]
I was just thinking about implementing this feature for some code I'm working on, if it's not already implemented.

**Comment by Dushistov** [comment]
May be generate const array or const function that return array that contains
all values?

Then implementation of iteration will be strait forward,
plus the array may also contains names:

```Rust
const fn all_values() -> [(Flag, &'static str); N]
```

This helps in implementation of custom serialization/deserialization (#147) or custom debug (#168).

What do you think?

At no I want parse / generate strings like: "AB BC CD" <-> `Flags::Ab | Flags::Bc | Flags::Cd`,

to generate string I need iteration plus names of constants in upper-case,
to deserialize I need names in upper-case

**Comment by sam0x17** [comment]
Would love this

**Comment by KodrAus** [comment]
In #278 we merged an implementation that allows iterating over flags and their names. It effectively looks like:

```rust
pub fn iter(self) -> impl Iterator<Item = (&'static str, Self)>
```

It effectively does what @Dushistov described, but internally. There's still design work to be done on this though, `IntoIterator` to implement, and some refactoring of how we generate flags types so we can make that `iter` function return a concrete type instead of `impl Trait` so I'll keep this ticket open for now to consider those.

**Comment by KodrAus** [comment]
This is now implemented in `2.0.0-rc.1`
